videoroom/6_ImplementingServerRoom.md

```diff
@@ -0,0 +1,164 @@
+
+ Here comes the next method:
+ ```elixir
+ @impl true
+ def handle_info({_sfu_engine, {:sfu_media_event, to, event}}, state) do
+    if state.peer_channels[to] != nil do
+        send(state.peer_channels[to], {:media_event, event})
+    end
+
+    {:noreply, state}
+ end
+ ```
+ The idea here is very similar to the one in the code snippet described previously - we want to direct the messages sent by SFU Engine's server to the SFU Engine's client.
+ The only difference is that not the event is about to be sent to a particular user - that is why instead of ```:broadcast``` atom as the second element of event's tuple we have ```to``` - which is a peer unique id. Since we precisely know to who we should send the message there is nothing else to do than to find the peer channel's process id associated with the given peer id (we are holding the (peer_id -> peer_channel_pid) mapping in the state of the GenServer!) and to send the message there. Once again the state does not need to change.
+
+
+ There we go with another message sent by SFU engine:
+ ```elixir
+ @impl true
+ def handle_info({sfu_engine, {:new_peer, peer_id, _metadata}}, state) do
+    # get node the peer with peer_id is running on
+    peer_channel_pid = Map.get(state.peer_channels, peer_id)
+    peer_node = node(peer_channel_pid)
+    send(sfu_engine, {:accept_new_peer, peer_id, peer_node})
+    {:noreply, state}
+ end
+ ```
+ That one might seem a little bit tricky. What is the deal here? Be aware that it is our process (based on Videoroom.Room module) who is the only one holding the mapping (peer_id->peer_channel_id). Once a new peer joins, the SFU Engine is not aware of any mapping between ```peer_id```and ```peer_channel_pid```. That is why SFU Engine, which is only aware of ```peer_id``` is asking our room process to give him some information about new peer - especially, the ```Kernel.node``` it belongs to (notice that due to use of BEAM virtual machine our application can be distributed - and server can be put on many machines working in the same cluster). To retrieve the information about the node on which peer channel exists, we need to refer to the process id of the peer channel process - and it is the room process that is aware of this process id.
```

Comment:
This part is really tricky and tbh I don't fully understand it yet :D Maybe we can make it clearer?

Reply:
Definitely, this part was a little bit messy - I have rewritten it in a clearer manner (or at least I hop so :D )
done

Reply:
test

Reply:
Indeed, this part was a little bit messy - I have rewritten it in a cleaner manner (or at least I hope so)